lens-cube/src/main/java/org/apache/lens/cube/parse/StorageTableResolver.java

```diff
@@ -245,85 +232,109 @@ private void resolveDimStorageTablesAndPartitions(CubeQueryContext cubeql) throw
+                                            : Arrays.asList(StringUtils.split(str.toLowerCase(), ","));
+      // Check if storagetable is in the list of valid storages.
+      if (validFactStorageTables != null && !validFactStorageTables.contains(storageTable)) {
+        log.info("Skipping storage table {} as it is not valid", storageTable);
+        cubeql.addStoragePruningMsg(sc, new CandidateTablePruneCause(CandidateTablePruneCode.INVALID_STORAGE));
+        it.remove();
         continue;
       }
 
-      Map<UpdatePeriod, Set<String>> storageTableMap = new TreeMap<UpdatePeriod, Set<String>>();
-      validStorageMap.put(fact, storageTableMap);
-      String str = conf.get(CubeQueryConfUtil.getValidStorageTablesKey(fact.getName()));
-      List<String> validFactStorageTables =
-        StringUtils.isBlank(str) ? null : Arrays.asList(StringUtils.split(str.toLowerCase(), ","));
-      Map<String, SkipStorageCause> skipStorageCauses = new HashMap<>();
-
-      for (Map.Entry<String, Set<UpdatePeriod>> entry : fact.getUpdatePeriods().entrySet()) {
-        String storage = entry.getKey();
-        // skip storages that are not supported
-        if (!isStorageSupported(storage)) {
-          log.info("Skipping storage: {} as it is not supported", storage);
-          skipStorageCauses.put(storage, new SkipStorageCause(SkipStorageCode.UNSUPPORTED));
-          continue;
+      boolean valid = false;
+      Set<CandidateTablePruneCause.CandidateTablePruneCode> codes = new HashSet<>();
+      for (TimeRange range : cubeql.getTimeRanges()) {
+        boolean columnInRange = client
+          .isStorageTableCandidateForRange(storageTable, range.getFromDate(), range.getToDate());
+        boolean partitionColumnExists = client.partColExists(storageTable, range.getPartitionColumn());
+        valid = columnInRange && partitionColumnExists;
+        if (valid) {
+          break;
         }
-        String table = getStorageTableName(fact, storage, validFactStorageTables);
-        // skip the update period if the storage is not valid
-        if (table == null) {
-          skipStorageCauses.put(storage, new SkipStorageCause(SkipStorageCode.INVALID));
+        if (!columnInRange) {
+          codes.add(TIME_RANGE_NOT_ANSWERABLE);
           continue;
         }
-        List<String> validUpdatePeriods =
-          CubeQueryConfUtil.getStringList(conf, CubeQueryConfUtil.getValidUpdatePeriodsKey(fact.getName(), storage));
-
-        boolean isStorageAdded = false;
-        Map<String, SkipUpdatePeriodCode> skipUpdatePeriodCauses = new HashMap<String, SkipUpdatePeriodCode>();
-        for (UpdatePeriod updatePeriod : entry.getValue()) {
-          if (maxInterval != null && updatePeriod.compareTo(maxInterval) > 0) {
-            log.info("Skipping update period {} for fact {}", updatePeriod, fact);
-            skipUpdatePeriodCauses.put(updatePeriod.toString(), SkipUpdatePeriodCode.QUERY_INTERVAL_BIGGER);
+        // This means fallback is required.
```

Comment:
I think the language should we "try fallback partition column" . It can onle be a try since there may be no fallback configured